### PR TITLE
[Merged by Bors] - fix(frontends/lean/elaborator): elaborate structure instances left-to-right

### LINKED ISSUE #197: projections unfolding while defining structures
We've seen a few instances of this general problem, in which while defining a structure, Lean unfolds too many projections (often breaking typeclass notations).

(In fact, today this bug is getting in the way of two separate projects of mine!)

This is an example provided by @urkud from https://leanprover.zulipchat.com/#narrow/stream/113488-general/topic/Lean.20unfolds.20projections.

```lean
import data.set.lattice

open set

namespace hidden

universes u v

variables (M : Type u) (G : Type v) [monoid M] [group G]

structure submonoid :=
(carrier : set M)
(one_mem' : (1:M) ∈ carrier)
(mul_mem' : ∀ (x ∈ carrier) (y ∈ carrier), x * y ∈ carrier)

structure subgroup extends submonoid G :=
(inv_mem' : ∀ x ∈ carrier, x⁻¹ ∈ carrier)

variables {M G}

instance : has_coe (subgroup G) (set G) := ⟨λ S, S.carrier⟩

instance : has_mem G (subgroup G) := ⟨λ x S, x ∈ (S : set G)⟩

instance example_1 : has_Inf (subgroup G) :=
⟨λ S, { carrier := ⋂ s ∈ S, ↑s,
  one_mem' := mem_bInter $ λ s hs, s.one_mem',
  mul_mem' := by { simp only [mem_bInter_iff], intros x hx y hy s hs, exact s.to_submonoid.mul_mem' _ (hx _ hs) _ (hy _ hs) },
  -- The goal for `inv_mem'` becomes:
  -- ⊢ ∀ (x : G), set.mem x (⋂ (s : subgroup G) (H : s ∈ S), ↑s) → set.mem x⁻¹ (⋂ (s : subgroup G) (H : s ∈ S), ↑s)
  -- Note in particular that `∈` has been unfolded inappropriately to `set.mem`.
  inv_mem' := by { trace_state, sorry } }⟩

-- One workaround is to introduce the variables in `mul_mem'`
-- before the tactic block.
instance example_2 : has_Inf (subgroup G) :=
⟨λ S, { carrier := ⋂ s ∈ S, ↑s,
  one_mem' := mem_bInter $ λ s hs, s.one_mem',
  mul_mem' := λ x hx y hy,
  begin
    simp only [mem_bInter_iff] at hx hy ⊢,
    intros s hs,
    exact s.to_submonoid.mul_mem' _ (hx _ hs) _ (hy _ hs)
  end,
  -- Now the goal contains a `{ carrier := ... }.carrier`,
  -- but even when we `dsimp` this, the `∈` is not disturbed.
  inv_mem' := by { dsimp, trace_state, sorry } }⟩

end hidden
```

### PR DESCRIPTION
**Pull request by gebner**
Fixes #197 and #214.

### CODE REVIEW

File: library/init/data/list/instances.lean [hunk 1]
```diff
@@ -16,10 +16,10 @@ instance : monad list :=
 { pure := @list.ret, map := @list.map, bind := @list.bind }
 
 instance : is_lawful_monad list :=
-{ bind_pure_comp_eq_map := by intros; induction x; simp [*, (<$>), pure] at *,
+{ bind_pure_comp_eq_map := by intros; induction x; simp [*, (<$>), (>>=), pure] at *,
```
Comment (gebner) [inline library/init/data/list/instances.lean:19]: The default value is less reduced now.

File: library/init/data/set.lean [hunk 2]
```diff
@@ -90,17 +90,9 @@ instance : functor set :=
...
-    exact propext ⟨λ ⟨b', ⟨_, _⟩⟩, ‹b' = b› ▸ ‹s b'›,
-                   λ _, ⟨b, ⟨‹s b›, rfl⟩⟩⟩,
-  end,
-  comp_map := begin
-    intros, funext c,
-    dsimp [image, set_of, function.comp],
-    exact propext ⟨λ ⟨a, ⟨h₁, h₂⟩⟩, ⟨g a, ⟨⟨a, ⟨h₁, rfl⟩⟩, h₂⟩⟩,
-                   λ ⟨b, ⟨⟨a, ⟨h₁, h₂⟩⟩, h₃⟩⟩, ⟨a, ⟨h₁, h₂.symm ▸ h₃⟩⟩⟩
-  end }
+{ id_map := λ _ s, funext $ λ b, propext ⟨λ ⟨_, sb, rfl⟩, sb, λ sb, ⟨_, sb, rfl⟩⟩,
```
Comment (gebner) [inline library/init/data/set.lean:93]: Same issue as in `list.monad`, and I was in a casual golfing mood.

File: tests/lean/t10.lean.expected.out [hunk 3]
```diff
@@ -1,4 +1,4 @@
-if p ∧ q then f x else y : N
+ifp ∧ qthenf xelsey : N
```
Comment (gebner) [inline tests/lean/t10.lean.expected.out:1]: @EdAyers Did you change something here?  I'm surprised the tests don't fail on master.

Comment (EdAyers) [reply tests/lean/t10.lean.expected.out:1]: I haven't changed this test. How curious.

Comment (EdAyers) [reply tests/lean/t10.lean.expected.out:1]: Don't we expect to have the spaces?

Comment (gebner) [reply tests/lean/t10.lean.expected.out:1]: Yes, I would definitely expect spaces to be here.

Comment (gebner) [reply tests/lean/t10.lean.expected.out:1]: And I'm aware you didn't change the test.  But you did change the pretty-printer.

**Comment by gebner** [comment]
I tried building mathlib and found something.  This change has a somewhat unintended side-effect.

For example, the `le_refl` goal now looks like this:
```lean
∀ a : α, @has_le.le (has_le.mk α (partial_order.le α)) a a
```

Unfortunately you can't `rw le_refl` here because the unifier fails to unify the two (at reducible transparency).  This is because projections of constructors are reduced somewhat eagerly and then you have a useless `partial_order.le`.

**Comment by kckennylau** [comment]
Does this fix #214?

**Comment by gebner** [comment]
@kckennylau This would indeed fix #214 (modulo the `has_mul.mk` issue I mentioned in the other comment).

**Comment by gebner** [comment]
So, I've ported mathlib and the diff isn't too large: https://github.com/leanprover-community/mathlib/compare/si

The workaround for the `has_mul.mk` issue above is as follows.  Consider for example `euclidean_domain.integral_domain`.  This would break:
```lean
@[priority 70] -- see Note [lower instance priority]
instance (α : Type*) [e : euclidean_domain α] : integral_domain α :=
by haveI := classical.dec_eq α; exact
{ eq_zero_or_eq_zero_of_mul_eq_zero :=
    λ a b (h : a * b = 0), or_iff_not_and_not.2 $ λ h0 : a ≠ 0 ∧ b ≠ 0,
      h0.1 $ by rw [← mul_div_cancel a h0.2, h, zero_div],
  ..e }
```

The reason is that the `..e` inserts a reference to `euclidean_domain.add`, which should never be referenced.  The workaround is very easy: just specify `mul := (*)` (etc.) manually:
```lean
@[priority 70] -- see Note [lower instance priority]
instance (α : Type*) [e : euclidean_domain α] : integral_domain α :=
by haveI := classical.dec_eq α; exact
{ eq_zero_or_eq_zero_of_mul_eq_zero := by exact
    λ a b h, (or_iff_not_and_not.2 $ λ h0,
      h0.1 $ by rw [← mul_div_cancel a h0.2, h, zero_div]),
  zero := 0, add := (+), mul := (*), ..e }
```

**Comment by gebner** [comment]
bors merge

**Pull request closed without merging**